atc/worker/pool.go

```diff
@@ -231,15 +232,27 @@ func (pool Pool) findWorkerForContainer(logger lager.Logger, owner db.ContainerO
 		return nil, nil, false, err
 	}
 
+	// When global-resources is enabled, a check-container may run on any team
+	// worker, which enables this optimization. Details refer to PR#8184.
+	if atc.EnableGlobalResources {
+		for _, w := range workersWithContainer {
+			if pool.isWorkerCompatibleAndRunning(logger, w, workerSpec) {
```

Comment:
Is there a way we can have it choose a team worker if it is available? And there isn't one, then we can use any worker?

I know some users use team workers to split up the workloads and it would be best if we can keep the same behaviour as before where it will prefer to use team workers if available and if not then it will use general workers.

Reply:
With this PR, it should never find more than 1 worker with target container. Do you agree?

With the example I used in the other comment. If resource-1 happens to run before the other two, then the check container will be placed on team-1's worker. When resource-2 runs, it will still reuse the check container from team-1's worker. If the team-1's worker is stalled, and resource-3 gets to run, then the check container will be recreated on team-3's worker. It should never find two containers.

Reply:
Right, I think this PR would make it choose the worker that has the target container on it, even if it is not from their team's worker. But I think before this PR, we would only choose the container if it was on one of the team workers. I'm just a little worried that the users will now see that they are running workloads on general workers rather than their own team workers. For example, previously when there is a team with team workers users would see that all their checks are run only on their team workers but after this pr, they might see their checks run on general workers. I do think global resources blurs this line with sharing of versions/check results, but this pr will take it a step further and actually run a team's resource checks on a non-team worker. We do define that team workers will take all workloads from that team in our docs https://concourse-ci.org/concourse-worker.html#team-workers so if we change this then we might want to add a note in there for users with team workers and global-resources.

Reply:
> I'm just a little worried that the users will now see that they are running workloads on general workers rather than their own team workers.

@clarafu I don't think that should happen. This PR doesn't modify the logic of initial check container creation, so if there is a team worker, a check container will be created on the team container.

What this PR changes is, a check container now may locate on other team's worker in case of the same scope. We can just enhance the doc to explain that.

Reply:
Are you sure? I don't think I was describing the initial check container creation, I think my example was explaining if there is a check container that exists for the scope on a general worker, before this change it would not use that check container and create a new container on a team worker. But after this change it will just use the check container on the general worker even if the team has a team worker. I'm just unsure about this fact because we give a guarantee that if you have team workers, all workloads will be scheduled onto those workers and we would be breaking that with this change.

Reply:
@clarafu I understood what you described, I just don't consider that's a problem. Let's use an example:

* teamA has resourceA and teamB has resourceB, and the two resources belong to the same scope
* both teams have no team worker
* a check container will be created on a general worker
* after some time, teamB got a team worker

Then our difference comes:

* My opinion is that, resourceB should still reuse the original check container
* Your opinion is that, resourceB should create a new check container on teamB's worker

Does that state clearly our difference?

As resourceB will still acquire the same resource-check-lock, share the same version history, and test the same scrope last check time as resourceA, what sense does it make to create a second check container on teamB's worker? If resourceA's check is always run a little bit earlier than resourceB, then due to the scope's last check time, resourceB might never get a chance to run check container.

So I think we should update the doc to explain that, when global resource is enabled, a check container might not strictly follow the team worker rule.

Reply:
According to what you describe above, i think the comment should be `a check-container may run on any worker` as long as the scope is same?